src/pybind/mgr/k8sevents/module.py

```diff
@@ -0,0 +1,1029 @@
+        changes.extend(self._check_osds(server_map, osd_map))
+        changes.extend(self._check_pools(pool_map))
+
+        # TODO
+        # Could generate an event if a ceph daemon has moved hosts
+        # (assumes the ceph metadata host information is valid - which may not be the case)
+
+        return changes
+
+    def run(self):
+        self.mgr.log.warning("[INF] Ceph configuration watcher started")
+
+        self.server_map, self.service_map = self.fetch_servers()
+        self.pool_map = self.fetch_pools()
+        # self.mgr.log.warning("[DBG] server map {}".format(self.server_map))
+        # self.mgr.log.warning("[DBG] service map {}".format(self.service_map))
+        self.osd_map = self.fetch_osd_map(self.service_map)
+        # self.mgr.log.warning("[DBG] osd map {}".format(self.osd_map))
+
+        ctr = 0
+        
+        while True:
+
+            try:
+                start_time = time.time()
+                server_map, service_map = self.fetch_servers()
+                pool_map = self.fetch_pools()
+                osd_map = self.fetch_osd_map(service_map)
+
+                changes = self.get_changes(server_map, osd_map, pool_map)
+                if changes:
+                    self.push_events(changes)
+
+                self.osd_map = osd_map
+                self.pool_map = pool_map
+                self.server_map = server_map
+                self.service_map = service_map
+
+                checks_duration = time.time() - start_time
+                if checks_duration > self.config_check_secs:
+                    new_interval = self.config_check_secs * 2
+                    self.mgr.log.warning("[WRN] k8sevents check interval warning. "
+                                         "Current checks {}s, interval is {}. "
+                                         "Increasing interval to {}".format(int(checks_duration),
+                                                                            self.config_check_secs,
+                                                                            new_interval))
+                    self.config_check_secs = new_interval
+
+                time.sleep(self.config_check_secs)    
+                ctr += self.config_check_secs
+                if ctr%self.heartbeat_interval_secs == 0:
+                    ctr = 0
+                    self.send_hearbeat_msg()
+            except BaseException:
+                self.health = "[ERR] Unexpected {} Exception @ line {}".format(sys.exc_info()[0].__name__, 
+                                                                               sys.exc_info()[2].tb_lineno)
+                break
+
+        self.mgr.log.warning("[WRN] Ceph configuration watcher stopped")
+        if self.health != 'OK':
+            self.mgr.log.error(self.health)
+
+class Module(MgrModule):
+    COMMANDS = [
+        {
+            "cmd": "k8sevents namespace",
+            "desc": "Show all Kuberenetes events from the rook-ceph namespace",
```

Comment:
Are all events permanently stored in the module's memory indefinitely?

Reply:
the namepace watcher events dict prunes itself - but the events generated by the module itself are not currently pruned. What do you think woul dbe a suitable retention period 3 days (to cover a weekend)?

Reply:
A month?

Reply:
to get us started, I've made 7 days the default, that can be overridden. I'm only pruning events we send to kubernetes - kubernetes events are pruned automatically.